forem/forem
# Migrate profile data

## What type of PR is this? (check all applicable)


## Description

Useful context: v1 of this feature tries to lay all the groundwork for generalized profiles for other communities, while not breaking DEV. This means that some of the code here sets up compatibility methods, which will eventually be removed.

This PR migrated profile data from `User` to `Profile`.  The idea is not to switch everything immediately but instead to the following:

1. Set up synchronisation between `User` and `Profile`.
    1. All new users will automatically have a filled profile. 
    1. Changes to profile related columns will still be stored on `User` so we can easily roll back, but they will also propagate to profiles. This shouldn't be too much of a problem and I'm trying to keep updates to the necessary minimum, but still making sure @mstruve sees this.
2. A Rake task for backfilling profile data for all users.

## Related Tickets & Documents

Part of #6994.

## QA Instructions, Screenshots, Recordings

* Run `rails temporary:migrate_profiles` and check the backfilled profiles.
* Create a new user and see that they automatically have a filled profile.
* Make changes to a user and ensure they propagate to the profile.

## Added tests?


## Added to documentation?


## Are there any post deployment tasks we need to perform?

We need to run `rails temporary:migrate_profiles` at one point, but the code is set up so nothing relies on that being done at a specific point in time. If we don't feel like kickstarting the task for all users, what we really need to do is run the following for each user once:

```ruby
Profile.create(user: user, data: Profiles::ExtractData.call(user))
```
